File: mars/src/test/java/MoveBackTest.java
import org.junit.jupiter.api.*;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class MoveBackTest {

    private static Area area;

    @BeforeAll
    static void init() {
        Point A = new Point(10, 10, Direction.SOUTH);
        Point B = new Point(-10, 10, Direction.SOUTH);
        Point C = new Point(-10, -10, Direction.SOUTH);
        Point D = new Point(10, -10, Direction.SOUTH);
        area = new Area(A, B, C, D);
    }

    @BeforeEach
    void setUp() {
    }

    @AfterEach
    void tearDown() {
    }

    @Test
    @DisplayName("北 -> 后退")
    void test_MoveBack_S() {
        MoveBack moveBack = new MoveBack(area);
        Point point = new Point(1, 1, Direction.SOUTH);
        assertEquals(point.getY() - 1, moveBack.execute(point).getY());
    }

    @Test
    @DisplayName("南 -> 后退")
    void test_MoveBack_N() {
        MoveBack moveBack = new MoveBack(area);
        Point point = new Point(1, 1, Direction.NORTH);
        assertEquals(point.getY() + 1, moveBack.execute(point).getY());
    }

    @Test
    @DisplayName("东 -> 后退")
    void test_MoveBack_E() {
        MoveBack moveBack = new MoveBack(area);
        Point point = new Point(1, 1, Direction.EAST);
        assertEquals(point.getX() - 1, moveBack.execute(point).getX());
    }

    @Test
    @DisplayName("西 -> 后退")
    void test_MoveBack_W() {
        MoveBack moveBack = new MoveBack(area);
        Point point = new Point(1, 1, Direction.WEST);
        assertEquals(point.getX() + 1, moveBack.execute(point).getX());
    }

    @Test
    @DisplayName("超出探索区域")
    void test_MoveBack_E_more() {
        MoveBack moveBack = new MoveBack(area);
        Point point = new Point(-10, 10, Direction.EAST);
        assertThrows(RuntimeException.class,
                () -> assertEquals(point.getX() - 1, moveBack.execute(point).getX()));
    }
}

File: mars-java-kotlin/src/test/kotlin/MoveBackTest.kt
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

internal class MoveBackTest {

    // 延迟初始化属性与变量
    lateinit var area: Area

    @BeforeEach
    fun setUp() {

        val A = Point(10, 10, Direction.SOUTH)
        val B = Point(-10, 10, Direction.SOUTH)
        val C = Point(-10, -10, Direction.SOUTH)
        val D = Point(10, -10, Direction.SOUTH)

        area = Area(A, B, C, D)
    }

    @Test
    @DisplayName("后退指令")
    fun excute() {

        assertEquals(-1, MoveBack(area).excute(Point(0, 0, Direction.SOUTH)).y)
        assertEquals(-1, MoveBack(area).excute(Point(0, 0, Direction.EAST)).x)
        // 越界
        assertThrows<RuntimeException> { MoveBack(area).excute(Point(-10, -10, Direction.EAST)).x }

    }

}